File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> ngeMap = new HashMap<>();
        Stack<Integer> mono = new Stack<>();
         for (int i : nums2) {
            while (!mono.isEmpty() && mono.peek() < i) {
                ngeMap.put(mono.pop(), i);
            }
            mono.push(i);
        }
 while (!mono.isEmpty()) {
            ngeMap.put(mono.pop(), -1);
        }

        // Step 2: Build the result for nums1 based on the precomputed map
        int[] result = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            result[i] = ngeMap.get(nums1[i]);
        }
        
        return result;
        
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        result={}
        stack=[]
        for i in range(len(nums2)):
            curr=nums2[i]
            while stack  and curr>nums2[stack[-1]]:
                ele=stack.pop()
                result[nums2[ele]] = curr
            stack.append(i)
        while stack:
            ele_index = stack.pop()
            result[nums2[ele_index]] = -1
            
        return [result[n] for n in nums1]


        